controllers/resourcepipeline_controller.go

```diff
@@ -0,0 +1,490 @@
+			reconciling.Reason = meta.ProgressingWithRetryReason
+			conditions.Set(obj, reconciling)
+		}
+
+		// If not reconciling or stalled than mark Ready=True
+		if !conditions.IsReconciling(obj) && !conditions.IsStalled(obj) && retErr == nil {
+			conditions.MarkTrue(
+				obj,
+				meta.ReadyCondition,
+				meta.SucceededReason,
+				"Reconciliation success",
+			)
+		}
+		// Set status observed generation option if the component is stalled or ready.
+		if conditions.IsStalled(obj) || conditions.IsReady(obj) {
+			obj.Status.ObservedGeneration = obj.Generation
+		}
+
+		// Update the object.
+		if perr := patchHelper.Patch(ctx, obj); perr != nil {
+			retErr = errors.Join(retErr, perr)
+		}
+	}()
+
+	// if the snapshot name has not been generated then
+	// generate, patch the status and requeue
+	if obj.GetSnapshotName() == "" {
+		name, err := snapshot.GenerateSnapshotName(obj.GetName())
+		if err != nil {
+			return ctrl.Result{}, err
+		}
+		obj.Status.SnapshotName = name
+		return ctrl.Result{Requeue: true}, nil
+	}
+
+	// Remove any stale Ready condition, most likely False, set above. Its value
+	// is derived from the overall result of the reconciliation in the deferred
+	// block at the very end.
+	conditions.Delete(obj, meta.ReadyCondition)
+
+	return r.reconcile(ctx, obj)
+}
+
+func (r *ResourcePipelineReconciler) reconcile(
+	ctx context.Context,
+	obj *v1alpha1.ResourcePipeline,
+) (ctrl.Result, error) {
+	logger := log.FromContext(ctx).WithName("resource-pipeline-controller")
+	conditions.MarkTrue(
+		obj,
+		meta.ReconcilingCondition,
+		meta.ProgressingReason,
+		"Reconciliation in progress",
+	)
+
+	// get the component
+	cv, err := r.getComponentVersionAccess(ctx, obj)
+	if err != nil {
+		return ctrl.Result{}, fmt.Errorf("component version not found: %w", err)
+	}
+	defer cv.Close()
+
+	// get the resource
+	res, err := cv.GetResource(ocmmetav1.NewIdentity(obj.Spec.SourceRef.ResourceRef.Name))
+	if err != nil {
+		return ctrl.Result{}, fmt.Errorf("resource not found: %w", err)
+	}
+
+	// prepare VFS
+	dir, err := os.MkdirTemp("", "wasm-tmp-")
+	if err != nil {
+		return ctrl.Result{}, fmt.Errorf("failed to get temp folder: %w", err)
+	}
+	defer os.RemoveAll(dir)
+
+	tmpfs, err := projectionfs.New(osfs.New(), dir)
+	if err != nil {
+		os.Remove(dir)
+		return ctrl.Result{}, fmt.Errorf("failed to create temp folder: %w", err)
+	}
+
+	// download resource to VFS
+	_, _, err = dirtree.New().Download(common.NewPrinter(os.Stdout), res, "", tmpfs)
+	if err != nil {
+		return ctrl.Result{}, fmt.Errorf("failed to download resource: %w", err)
+	}
+
+	parameters := map[string]any{}
+	if obj.Spec.Parameters != nil {
+		if err := json.Unmarshal(obj.Spec.Parameters.Raw, &parameters); err != nil {
+			return ctrl.Result{}, fmt.Errorf("failed to parse parameters: %w", err)
+		}
+	}
+
+	for _, step := range obj.Spec.PipelineSpec.Steps {
+		if err := r.executePipelineStep(ctx, step, nil, cv, dir, parameters); err != nil {
```

Comment:
I'm wondering if we should embed the outside context, but otherwise create a new one with its own timeout for running these scripts to potentially cancel the run. 🤔

Reply:
that's a good call.

Reply:
nice!